### ISSUE TITLE
RFE: Document org.freedesktop.systemd1.Privileged DBus annotation

### ISSUE THREAD

**Issue by juspence** [systemd/systemd#21882]
**Is your feature request related to a problem? Please describe.**
I'm always frustrated when I try to call a DBus method but get "Rejected send message" errors, even though I already updated the DBus policy to allow this method call:
````
Dec 23 21:51:10 localhost dbus-daemon[5804]: [system] Rejected send message, 2 matched rules; type="method_call", sender=":1.27" (uid=42 pid=6292 comm="gdm-session-worker [pam/gdm-launch-environment] " label="system_u:system_r:unconfined_service_t:s0") interface="org.freedesktop.login1.Manager" member="CreateSession" error name="(unset)" requested_reply="0" destination="org.freedesktop.login1" (uid=0 pid=5812 comm="/usr/lib/systemd/systemd-logind " label="system_u:system_r:systemd_logind_t:s0")
````

The context here is that I'm trying to run GDM & agetty as unprivileged users instead of as root. This mostly works for agetty - a new VT is started the way I expect, I can log in and out with it, and I don't have any problems while running commands. But every time I log in I see the error message above in journal logs. GDM does not let me log in at all - this error message prevents GDM from starting a graphical session.

My custom DBus policy allows the gdm & agetty users to make any method call to org.freedesktop.login1:
````
$ cat /etc/dbus-1/system.d/org.freedesktop.login1.conf
<?xml version="1.0"?> <!--*-nxml-*-->
<!DOCTYPE busconfig PUBLIC "-//freedesktop//DTD D-BUS Bus Configuration 1.0//EN"
        "http://www.freedesktop.org/standards/dbus/1.0/busconfig.dtd">

<busconfig>
        <policy user="root">
                <allow own="org.freedesktop.login1"/>
                <allow send_destination="org.freedesktop.login1"/>
                <allow receive_sender="org.freedesktop.login1"/>
        </policy>
        <!-- Added by me to allow unprivileged agetty logins -->
        <policy user="agetty">
                <allow own="org.freedesktop.login1"/>
                <allow send_destination="org.freedesktop.login1"/>
                <allow receive_sender="org.freedesktop.login1"/>
                <allow send_destination="org.freedesktop.login1"
                       send_interface="org.freedesktop.login1.Manager"
                       send_member="CreateSession"/>
                <allow send_destination="org.freedesktop.login1"
                       send_interface="org.freedesktop.login1.Manager"
                       send_member="ReleaseSession"/>
                     https://freedesktop.org/wiki/Software/systemd/logind/
        </policy>
        <!-- Added by me to allow unprivileged gdm logins -->
        <policy user="gdm">
                <allow own="org.freedesktop.login1"/>
                <allow send_destination="org.freedesktop.login1"/>
                <allow receive_sender="org.freedesktop.login1"/>
                <allow send_destination="org.freedesktop.login1"
                       send_interface="org.freedesktop.login1.Manager"
                       send_member="CreateSession"/>
                <allow send_destination="org.freedesktop.login1"
                       send_interface="org.freedesktop.login1.Manager"
                       send_member="ReleaseSession"/>
        </policy>

        <policy context="default">
                <deny send_destination="org.freedesktop.login1"/>

                <allow send_destination="org.freedesktop.login1"
                       send_interface="org.freedesktop.DBus.Introspectable"/>
                <!-- Many more lines removed from only this output
                       they allow calling some other methods by default
                       but not CreateSession or ReleaseSession -->
</busconfig>
````

Trying to call CreateSession and ReleaseSession using dbus-send says "permission denied" when run as the gdm user. Permissions checks succeed (and the call still fails due to a missing argument, as expected) when run as root:
````
$ sudo -u gdm dbus-send --system --dest=org.freedesktop.login1 --type=method_call --print-reply /org/freedesktop/login1 org.freedesktop.login1.Manager.CreateSession
Error org.freedesktop.DBus.Error.AccessDenied: Sender is not authorized to send message
$ sudo dbus-send --system --dest=org.freedesktop.login1 --type=method_call --print-reply /org/freedesktop/login1 org.freedesktop.login1.Manager.CreateSession
Error org.freedesktop.DBus.Error.InvalidArgs: Invalid arguments '' to call org.freedesktop.login1.Manager.CreateSession(), expecting 'uusssssussbssa(sv)'.
````

I introspected the "org.freedesktop.login1" bus and saw that "CreateSession" and "ReleaseSession" have the "org.freedesktop.systemd1.Privileged" annotation:
````
$ sudo dbus-send --system --dest=org.freedesktop.login1 --type=method_call --print-reply /org/freedesktop/login1 org.freedesktop.DBus.Introspectable.Introspect | grep 'Privileged' --before-context=23 --after-context=1
  <method name="ReleaseSession">
   <arg type="s" name="session_id" direction="in"/>
   <annotation name="org.freedesktop.systemd1.Privileged" value="true"/>
  </method>
````

https://www.freedesktop.org/software/systemd/man/org.freedesktop.login1.html only mentions the "privileged" annotation for the "EnableWallMessages" and "WallMessage" property:
````
               properties:
                 @org.freedesktop.DBus.Property.EmitsChangedSignal("false")
                 @org.freedesktop.systemd1.Privileged("true")
                 readwrite b EnableWallMessages = ...;
                 @org.freedesktop.DBus.Property.EmitsChangedSignal("false")
                 @org.freedesktop.systemd1.Privileged("true")
                 readwrite s WallMessage = '...';
````

That man page also mentions "privileged" and "unprivileged" users. https://www.freedesktop.org/software/systemd/man/org.freedesktop.LogControl1.html also has two "privileged" properties, and explains: "Those two properties are writable, so they may be set by sufficiently privileged users". But there's no description of how this annotation works or what "sufficiently privileged users" means. Is only the root user considered privileged, or is there some way to grant privileges to a non-root user?

**Describe the solution you'd like**
Please document how the "org.freedesktop.systemd1.Privileged" annotation works, how a user can gain privileges (if possible), and which users are considered "privileged" versus "unprivileged" by default.  I think it would be helpful to add these to the org.freedesktop.login1 man page. Or more explicit notes / a separate man page could be created for how Systemd checks DBus method caller privileges.

I found a hint on the https://www.freedesktop.org/software/systemd/man/org.freedesktop.systemd1.html man page:
````
Security

Similarly to methods on the Manager object, read-only access is allowed for everyone. All operations are allowed for clients with the CAP_SYS_ADMIN capability or when the org.freedesktop.systemd1.manage-units privilege is granted by polkit.
````

But GDM fails to call "CreateSession" even when run with all capabilities (including CAP_SYS_ADMIN), and this doesn't agree with another hint in the https://www.freedesktop.org/software/systemd/man/systemd.journal-fields.html man page: "Privileged programs (currently UID 0) may attach /OBJECT_PID= to a message".

**Describe alternatives you've considered**
In Polkit, annotations like below can be added to authorize a (non-root) user to call CheckAuthorization on some action:
````
        <action id="org.freedesktop.resolve1.register-service">
                <description gettext-domain="systemd">Register a DNS-SD service</description>
                <message gettext-domain="systemd">Authentication is required to register a DNS-SD service</message>
                <defaults>
                        <allow_any>auth_admin</allow_any>
                        <allow_inactive>auth_admin</allow_inactive>
                        <allow_active>auth_admin_keep</allow_active>
                </defaults>
                <annotate key="org.freedesktop.policykit.owner">unix-user:systemd-resolve</annotate>
        </action>
````

Without this annotation, trying to use e.g. pkcheck will fail with a "permission denied" error message:
````
$ sudo -u gdm pkcheck -a 'org.freedesktop.resolve1.register-service' --process $(pgrep bash) -u
Error checking for authorization org.freedesktop.resolve1.register-service: GDBus.Error:org.freedesktop.PolicyKit1.Error.NotAuthorized: Only trusted callers (e.g. uid 0 or an action owner) can use CheckAuthorization() for subjects belonging to other identities
$ sudo -u systemd-resolve pkcheck -a 'org.freedesktop.resolve1.register-service' --process $(pgrep bash) -u
polkit\56retains_authorization_after_challenge=1
polkit\56retains_authorization_after_challenge=true
````

It would be nice if systemd supported something similar (maybe it already does?). Then I can add some config file which says the gdm and agetty users are trusted, and are allowed to call the "CreateSession" and "ReleaseSession" DBus methods on the "org.freedesktop.login1" bus.

If this is not desirable, and only the root user is considered privileged, maybe a code change can be made. Instead of assuming that only "UID 0" is trusted, systemd-logind could assume that only "the user matching my effective UID" is trusted.

The result will be the same when systemd-logind runs as root, but enables my usecase where I want to run gdm and agetty as non-root. If I create a new "login" user that GDM, agetty, and systemd-logind all share, then GDM and agetty should be able to call CreateSession without getting this error message.

**The systemd version you checked that didn't have the feature you are asking for**
N/A, it's a docs issue. Local and online (so latest) man pages don't explain this part of Systemd well.

**Comment by yuwata** [comment]
Fix for the document issue is waiting in #21844.

> It would be nice if systemd supported something similar (maybe it already does?). Then I can add some config file which says the gdm and agetty users are trusted, and are allowed to call the "CreateSession" and "ReleaseSession" DBus methods on the "org.freedesktop.login1" bus.

> If this is not desirable, and only the root user is considered privileged, maybe a code change can be made. Instead of assuming that only "UID 0" is trusted, systemd-logind could assume that only "the user matching my effective UID" is trusted.

> The result will be the same when systemd-logind runs as root, but enables my usecase where I want to run gdm and agetty as non-root. If I create a new "login" user that GDM, agetty, and systemd-logind all share, then GDM and agetty should be able to call CreateSession without getting this error message.

Not sure about the suggestion. If you edit the policy, then does it work for you? Or, we need to update some code to support such use case?

**Comment by juspence** [comment]
@yuwata Thanks for the PR, but this didn't fix my issue. I would like to know:

1) How do "privileged DBus methods" work? What permissions are needed to call a privileged DBus method?
2) How do "privileged users" worK? Which users are privileged - only root?
3) Can I grant privileges to some daemon user? What config file do I edit?

> Not sure about the suggestion. If you edit the policy, then does it work for you? Or, we need to update some code to support such use case?

I looked at the "org.freedesktop.login1.policy" Polkit file, but it has no "org.freedesktop.login1.CreateSession" action. Is this the policy you are suggesting I edit? Or is there a different policy file for these privileged DBus methods?

@bluca Given the above, can you please reopen this?
